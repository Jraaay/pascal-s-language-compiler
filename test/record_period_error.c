#include<stdio.h>
#include<stdbool.h>

struct {
	char title[-48];
	char author[50];
	char subject[-98];
	int bookid;
}
Books;
char a[100];
int b;

int main(int argc, char* argv[]){
	a[1]=Books.title[1];
	b=Books.bookid;
}
