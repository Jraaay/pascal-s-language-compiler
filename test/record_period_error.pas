{Bonus Features: Record}

Program example(input, output);


var
Books : record
  title: array [50..1] of char;
  author: array [1..50] of char;
  subject: array [100..1] of char;
  bookid: integer;
end;
a : array [1..100] of char;
b : integer;

Begin
  a[1] := Books.title[1];
  b := Books.bookid;
End.
