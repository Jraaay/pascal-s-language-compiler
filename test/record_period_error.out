{
    "ast": {
        "type": "programstruct",
        "program_head": {
            "type": "program_head",
            "ID": "example",
            "idlist": {
                "type": "idlist",
                "ids": [
                    "input",
                    "output"
                ]
            }
        },
        "program_body": {
            "type": "program_body",
            "const_declarations": null,
            "var_declarations": {
                "type": "var_declarations",
                "var_declaration": {
                    "type": "var_declaration",
                    "values": [
                        {
                            "idlist": {
                                "type": "idlist",
                                "ids": [
                                    "Books"
                                ]
                            },
                            "type": {
                                "type": "type",
                                "_type": "RECORD",
                                "multype": {
                                    "type": [
                                        {
                                            "type": "type",
                                            "_type": "ARRAY",
                                            "period": {
                                                "type": "period",
                                                "values": [
                                                    {
                                                        "start": 50,
                                                        "END": 1
                                                    }
                                                ]
                                            },
                                            "basic_type": "of"
                                        },
                                        {
                                            "type": "type",
                                            "_type": "ARRAY",
                                            "period": {
                                                "type": "period",
                                                "values": [
                                                    {
                                                        "start": 1,
                                                        "END": 50
                                                    }
                                                ]
                                            },
                                            "basic_type": "of"
                                        },
                                        {
                                            "type": "type",
                                            "_type": "ARRAY",
                                            "period": {
                                                "type": "period",
                                                "values": [
                                                    {
                                                        "start": 100,
                                                        "END": 1
                                                    }
                                                ]
                                            },
                                            "basic_type": "of"
                                        },
                                        {
                                            "type": "type",
                                            "_type": {
                                                "type": "basic_type",
                                                "_type": "integer"
                                            }
                                        }
                                    ],
                                    "multype": {
                                        "type": [
                                            {
                                                "type": "type",
                                                "_type": "ARRAY",
                                                "period": {
                                                    "type": "period",
                                                    "values": [
                                                        {
                                                            "start": 50,
                                                            "END": 1
                                                        }
                                                    ]
                                                },
                                                "basic_type": "of"
                                            },
                                            {
                                                "type": "type",
                                                "_type": "ARRAY",
                                                "period": {
                                                    "type": "period",
                                                    "values": [
                                                        {
                                                            "start": 1,
                                                            "END": 50
                                                        }
                                                    ]
                                                },
                                                "basic_type": "of"
                                            },
                                            {
                                                "type": "type",
                                                "_type": "ARRAY",
                                                "period": {
                                                    "type": "period",
                                                    "values": [
                                                        {
                                                            "start": 100,
                                                            "END": 1
                                                        }
                                                    ]
                                                },
                                                "basic_type": "of"
                                            }
                                        ],
                                        "multype": {
                                            "type": [
                                                {
                                                    "type": "type",
                                                    "_type": "ARRAY",
                                                    "period": {
                                                        "type": "period",
                                                        "values": [
                                                            {
                                                                "start": 50,
                                                                "END": 1
                                                            }
                                                        ]
                                                    },
                                                    "basic_type": "of"
                                                },
                                                {
                                                    "type": "type",
                                                    "_type": "ARRAY",
                                                    "period": {
                                                        "type": "period",
                                                        "values": [
                                                            {
                                                                "start": 1,
                                                                "END": 50
                                                            }
                                                        ]
                                                    },
                                                    "basic_type": "of"
                                                }
                                            ],
                                            "multype": {
                                                "type": [
                                                    {
                                                        "type": "type",
                                                        "_type": "ARRAY",
                                                        "period": {
                                                            "type": "period",
                                                            "values": [
                                                                {
                                                                    "start": 50,
                                                                    "END": 1
                                                                }
                                                            ]
                                                        },
                                                        "basic_type": "of"
                                                    }
                                                ],
                                                "ID": [
                                                    "title"
                                                ]
                                            },
                                            "ID": [
                                                "title",
                                                "author"
                                            ]
                                        },
                                        "ID": [
                                            "title",
                                            "author",
                                            "subject"
                                        ]
                                    },
                                    "ID": [
                                        "title",
                                        "author",
                                        "subject",
                                        "bookid"
                                    ]
                                }
                            }
                        },
                        {
                            "idlist": {
                                "type": "idlist",
                                "ids": [
                                    "a"
                                ]
                            },
                            "type": {
                                "type": "type",
                                "_type": "ARRAY",
                                "period": {
                                    "type": "period",
                                    "values": [
                                        {
                                            "start": 1,
                                            "END": 100
                                        }
                                    ]
                                },
                                "basic_type": "of"
                            }
                        },
                        {
                            "idlist": {
                                "type": "idlist",
                                "ids": [
                                    "b"
                                ]
                            },
                            "type": {
                                "type": "type",
                                "_type": {
                                    "type": "basic_type",
                                    "_type": "integer"
                                }
                            }
                        }
                    ]
                }
            },
            "subprogram_declarations": {
                "type": "subprogram_declarations",
                "subprograms": []
            },
            "compound_statement": {
                "type": "compound_statement",
                "statement_list": {
                    "type": "statement_list",
                    "statements": [
                        {
                            "type": "statement",
                            "_type": "variable",
                            "variable": {
                                "type": "variable",
                                "ID": "a",
                                "id_varpart": {
                                    "type": "id_varpart",
                                    "expression_list": {
                                        "type": "expression_list",
                                        "expressions": [
                                            {
                                                "type": "expression",
                                                "simple_expression": {
                                                    "type": "simple_expression",
                                                    "term": {
                                                        "type": "term",
                                                        "factor": {
                                                            "type": "factor",
                                                            "_type": "NUM",
                                                            "NUM": 1
                                                        }
                                                    }
                                                }
                                            }
                                        ]
                                    }
                                }
                            },
                            "ASSIGNOP": ":=",
                            "expression": {
                                "type": "expression",
                                "simple_expression": {
                                    "type": "simple_expression",
                                    "term": {
                                        "type": "term",
                                        "factor": {
                                            "type": "factor",
                                            "_type": "variable",
                                            "variable": {
                                                "type": "variable",
                                                "ID": [
                                                    "Books",
                                                    "title"
                                                ],
                                                "id_varpart": {
                                                    "type": "id_varpart",
                                                    "expression_list": {
                                                        "type": "expression_list",
                                                        "expressions": [
                                                            {
                                                                "type": "expression",
                                                                "simple_expression": {
                                                                    "type": "simple_expression",
                                                                    "term": {
                                                                        "type": "term",
                                                                        "factor": {
                                                                            "type": "factor",
                                                                            "_type": "NUM",
                                                                            "NUM": 1
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        ]
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        },
                        {
                            "type": "statement",
                            "_type": "variable",
                            "variable": {
                                "type": "variable",
                                "ID": "b",
                                "id_varpart": {
                                    "type": "id_varpart",
                                    "expression_list": null
                                }
                            },
                            "ASSIGNOP": ":=",
                            "expression": {
                                "type": "expression",
                                "simple_expression": {
                                    "type": "simple_expression",
                                    "term": {
                                        "type": "term",
                                        "factor": {
                                            "type": "factor",
                                            "_type": "variable",
                                            "variable": {
                                                "type": "variable",
                                                "ID": [
                                                    "Books",
                                                    "bookid"
                                                ],
                                                "id_varpart": {
                                                    "type": "id_varpart",
                                                    "expression_list": null
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    ]
                }
            }
        }
    },
    "symbolTable": {
        "constants": [],
        "variables": [
            {
                "id": 4,
                "token": "Books",
                "type": "RECORD",
                "isArray": false,
                "dimension": 0,
                "size": [],
                "start": [],
                "recordTable": {
                    "variables": [
                        {
                            "id": 0,
                            "token": "title",
                            "type": "char",
                            "isArray": true,
                            "dimension": 1,
                            "size": [
                                -48
                            ],
                            "start": [
                                50
                            ],
                            "recordTable": null
                        },
                        {
                            "id": 1,
                            "token": "author",
                            "type": "char",
                            "isArray": true,
                            "dimension": 1,
                            "size": [
                                50
                            ],
                            "start": [
                                1
                            ],
                            "recordTable": null
                        },
                        {
                            "id": 2,
                            "token": "subject",
                            "type": "char",
                            "isArray": true,
                            "dimension": 1,
                            "size": [
                                -98
                            ],
                            "start": [
                                100
                            ],
                            "recordTable": null
                        },
                        {
                            "id": 3,
                            "token": "bookid",
                            "type": "integer",
                            "isArray": false,
                            "dimension": 0,
                            "size": [],
                            "start": [],
                            "recordTable": null
                        }
                    ]
                }
            },
            {
                "id": 5,
                "token": "a",
                "type": "char",
                "isArray": true,
                "dimension": 1,
                "size": [
                    100
                ],
                "start": [
                    1
                ],
                "recordTable": null
            },
            {
                "id": 6,
                "token": "b",
                "type": "integer",
                "isArray": false,
                "dimension": 0,
                "size": [],
                "start": [],
                "recordTable": null
            }
        ],
        "subFunc": []
    },
    "error": [
        {
            "code": "C-01",
            "info": {
                "line": 5,
                "value": [
                    50,
                    1
                ],
                "lexpos": 101
            }
        },
        {
            "code": "C-01",
            "info": {
                "line": 7,
                "value": [
                    100,
                    1
                ],
                "lexpos": 169
            }
        }
    ]
}